### PR TITLE
Autotester Token Generation Time

### LINKED ISSUE #7071: Display when the next "Student-Run Tests" token generation happens
We want to give a specific time in the future when the next "Student-Run Tests" token will be generated. This will be helpful for students to know when to expect the next token generation to happen.

Currently it shows the following:
image

### PR DESCRIPTION
**Pull request by umututku03**
## Proposed Changes
*(Describe your changes here. Also describe the motivation for your changes: what problem do they solve, or how do they improve the application or codebase? If this pull request fixes an open issue, [use a keyword to link this pull request to the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).)*

This enhancement aims to help students know exactly when they can expect the next token to be available, providing clarity and better planning for their test runs. Currently, the system shows - tokens available on, tokens regenerate after and token available. The goal is to calculate and display the exact future time when the next token will be generated based on the current token regeneration period and the last token generation time. We refactored the `AutomatedTestsController` to improve error handling and robustly calculate the `next_token_generation_time`, added new I18n keys for better user messaging, and ensured all keys are used correctly to avoid non-normalized keys. The frontend interface was updated to display these messages clearly, informing students about next token generation time.

Resolves #7071.

<details>
<summary>Screenshots of your changes (if applicable)</summary>
Screenshot 2024-07-12 at 6 53 23 PM

</details>

<details>
<summary>Associated <a href="https://github.com/MarkUsProject/Wiki">documentation repository</a> pull request (if applicable)</summary>

</details>

## Type of Change
*(Write an `X` or a brief description next to the type or types that best describe your changes.)*

| Type                                                                                    | Applies? |
|-----------------------------------------------------------------------------------------|----------|
| 🚨 *Breaking change* (fix or feature that would cause existing functionality to change) |    X      |
| ✨ *New feature* (non-breaking change that adds functionality)                          |      X    |
| 🐛 *Bug fix* (non-breaking change that fixes an issue)                                  |          |
| 🎨 *User interface change* (change to user interface; provide screenshots)              |      X    |
| ♻️ *Refactoring* (internal change to codebase, without changing functionality)          |          |
| 🚦 *Test update* (change that *only* adds or modifies tests)                            |          |
| 📦 *Dependency update* (change that updates a dependency)                               |          |
| 🔧 *Internal* (change that *only* affects developers or continuous integration)         |          |

## Checklist
*(Complete each of the following items for your pull request. Indicate that you have completed an item by changing the `[ ]` into a `[x]` in the raw text, or by clicking on the checkbox in the rendered description on GitHub.)*

Before opening your pull request:

- [X] I have performed a self-review of my changes.
  - Check that all changed files included in this pull request are intentional changes.
  - Check that all changes are relevant to the purpose of this pull request, as described above.
- [X] I have added tests for my changes, if applicable.
  - This is **required** for all bug fixes and new features.
- [ ] I have updated the project documentation, if applicable.
  - This is **required** for new features.
- [x] If this is my first contribution, I have added myself to the list of contributors.

After opening your pull request:

- [x] I have updated the project Changelog (this is required for all changes).
- [X] I have verified that the pre-commit.ci checks have passed.
- [x] I have verified that the CI tests have passed.
- [ ] I have reviewed the test coverage changes reported by Coveralls.
- [ ] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments
*(Include any questions or comments you have regarding your changes.)*

### CODE REVIEW

File: app/models/assignment.rb [hunk 1]
```diff
@@ -24,6 +24,8 @@ class Assignment < Assessment
   before_save :reset_collection_time
   after_create :update_parent_assignment, if: :is_peer_review?
 
+  delegate :enable_student_tests, :token_period, to: :assignment_properties
```
Comment (david-yz-liu) [inline app/models/assignment.rb:27]: This line isn't necessary, as the `delegate_missing_to` above will handle these two attributes

File: app/controllers/automated_tests_controller.rb [hunk 2]
```diff
@@ -35,12 +35,26 @@ def student_interface
...
+      if @assignment.enable_student_tests
+        # Calculate the next token generation time
+        token_period = @assignment.token_period
+        last_generated = @grouping.test_runs.first&.created_at || Time.zone.now
+        @next_token_generation_time = last_generated + token_period.hours
+
+        # Format the next token generation time for display
+        @next_token_generation_time = @next_token_generation_time.strftime('%A, %B %d, %Y %I:%M:%S %p %Z')
+      else
+        @next_token_generation_time = nil
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:56]: Combine this case with the original condition on line 38.

File: app/controllers/automated_tests_controller.rb [hunk 3]
```diff
@@ -35,12 +35,26 @@ def student_interface
...
                                                  context: { assignment: @assignment, grouping: @grouping })).value
+
+      if @assignment.enable_student_tests
+        # Calculate the next token generation time
+        token_period = @assignment.token_period
+        last_generated = @grouping.test_runs.first&.created_at || Time.zone.now
+        @next_token_generation_time = last_generated + token_period.hours
+
+        # Format the next token generation time for display
+        @next_token_generation_time = @next_token_generation_time.strftime('%A, %B %d, %Y %I:%M:%S %p %Z')
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:54]: You can use `I18n.l` in controllers to do the same type of date formatting as in views

File: app/controllers/automated_tests_controller.rb [hunk 4]
```diff
@@ -35,12 +35,26 @@ def student_interface
...
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
+
+      if @assignment.enable_student_tests
+        # Calculate the next token generation time
+        token_period = @assignment.token_period
+        last_generated = @grouping.test_runs.first&.created_at || Time.zone.now
+        @next_token_generation_time = last_generated + token_period.hours
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:51]: This isn't the correct logic for token refresh time. Look at `Grouping#refresh_test_tokens`

File: app/views/automated_tests/student_interface.html.erb [hunk 5]
```diff
@@ -4,15 +4,15 @@
 <%= javascript_include_tag 'result', nonce: true %>
 <%= javascript_include_tag 'Results/main', nonce: true %>
 
-<% if @assignment.enable_student_tests && !@grouping.nil? %>
+<% if @assignment.enable_student_tests && @grouping.present? %>
```
Comment (david-yz-liu) [inline app/views/automated_tests/student_interface.html.erb:7]: Revert this change (and the one below) since it's in a separate pull request. Note that you can modify the token end date in the automated tests settings tab for testing purposes.

File: spec/controllers/automated_tests_controller_spec.rb [hunk 6]
```diff
@@ -1,7 +1,8 @@
+require 'rails_helper'
```
Comment (david-yz-liu) [inline spec/controllers/automated_tests_controller_spec.rb:1]: This isn't necessary to include

File: spec/controllers/automated_tests_controller_spec.rb [hunk 7]
```diff
@@ -1,7 +1,8 @@
...
+
 describe AutomatedTestsController do
   include AutomatedTestsHelper
 
-  # TODO: add 'role is from a different course' shared tests to each route test below
```
Comment (david-yz-liu) [inline spec/controllers/automated_tests_controller_spec.rb:4]: You have not resolved this TODO, which is fine but you shouldn't delete it

File: spec/controllers/automated_tests_controller_spec.rb [hunk 8]
```diff
@@ -548,8 +549,79 @@
...
+                 token_period: 24
+               })
+      end
+      let(:grouping) { create(:grouping, assignment: assignment) }
+      let(:student) { create(:student) }
+      let(:role) { student }
+      let(:params) { { course_id: assignment.course.id, assignment_id: assignment.id } }
+
+      before do
+        allow_any_instance_of(AutotestSetting).to(
```
Comment (david-yz-liu) [inline spec/controllers/automated_tests_controller_spec.rb:567]: This isn't necessary to include for these tests

File: spec/controllers/automated_tests_controller_spec.rb [hunk 9]
```diff
@@ -548,8 +549,79 @@
...
+        sign_in student
+      end
+
+      context 'when student tests are enabled and the student has a grouping' do
+        before do
+          get_as student, :student_interface, params: params
+        end
+
+        it 'should calculate the next token generation time' do
+          expect(assigns(:next_token_generation_time)).not_to be_nil
```
Comment (david-yz-liu) [inline spec/controllers/automated_tests_controller_spec.rb:580]: This test should be more specific. Calculate the exact time that should be displayed. This is related to my above comment about this time should actually be generated.

**Review by david-yz-liu — COMMENTED** [review]
@umututku03 in addition to my inline comments, you should redo the pull request description, as you have removed parts under the "description" section and have some formatting errors in the checklist. Note that since this is a user-facing change screenshots should be included.

Also, there is a failing test.

File: app/controllers/automated_tests_controller.rb [hunk 10]
```diff
@@ -35,14 +35,34 @@ def student_interface
...
+    if @grouping.present?
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
+
+      if @assignment.enable_student_tests
+        last_student_run = @grouping.test_runs.where(role: @grouping.accepted_students).first
+        if last_student_run.nil?
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:47]: This check is not necessary, the only logic that's needed is in the else branch. If the student hasn't run any tests, we still need to take into account the current time, as that might be very far past when the `@assignment.token_start_date` is.

File: app/controllers/automated_tests_controller.rb [hunk 11]
```diff
@@ -35,14 +35,34 @@ def student_interface
...
+          @next_token_generation_time = @assignment.token_start_date + @assignment.token_period.hours
+        else
+          hours_from_start = (Time.current - @assignment.token_start_date) / 3600
+          if @assignment.non_regenerating_tokens
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:51]: This is a good case to consider, but should actually be handled separately. When tokens do not regenerate, or when there are *unlimited tokens*, we should not calculate a value for `@next_token_generation_time` at all, but instead just set it to `nil`. This can be checked in the condition on line 45.

File: app/controllers/automated_tests_controller.rb [hunk 12]
```diff
@@ -35,14 +35,34 @@ def student_interface
...
+            last_period_begin = @assignment.token_start_date
+          else
+            periods_from_start = (hours_from_start / @assignment.token_period).floor
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:54]: This calculation is good, but doesn't correctly handle the case when the current time is *before* the assignment token start date. In that case, this value will be negative, but you should set it to `0` instead.

File: app/views/automated_tests/student_interface.html.erb [hunk 13]
```diff
@@ -12,7 +12,7 @@
 
   <p>
     <strong><%= Assignment.human_attribute_name(:token_end_date) %>:</strong>
-    <%= l(@assignment.token_end_date) %>
+    <%= @assignment.token_end_date.present? ? l(@assignment.token_end_date) : '-' %>
```
[the remark this answers is not in the document]
Comment (umututku03) [reply app/views/automated_tests/student_interface.html.erb:15]: I think this should not be reversed. If reversed, the tests fail due to an issue I mentioned (a few weeks ago) about 'nil' values.

File: app/controllers/automated_tests_controller.rb [hunk 14]
```diff
@@ -35,14 +35,26 @@ def student_interface
...
+      if @assignment.enable_student_tests
+        if @assignment.non_regenerating_tokens
+          @next_token_generation_time = nil
+        else
+          hours_from_start = [(Time.current - @assignment.token_start_date) / 3600, 0].max
+          periods_from_start = (hours_from_start / @assignment.token_period).floor
+          last_period_begin = @assignment.token_start_date + (periods_from_start * @assignment.token_period).hours
+          @next_token_generation_time = last_period_begin + @assignment.token_period.hours
+        end
+        @next_token_generation_time = I18n.l(@next_token_generation_time) if @next_token_generation_time
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:54]: This line of code should be merged into the else branch above, and you don't need the `if @next_token_generation_time`. That is, if the else branch executes then we need to call `I18n.l`, otherwise we don't.

File: app/controllers/automated_tests_controller.rb [hunk 15]
```diff
@@ -35,14 +35,26 @@ def student_interface
...
+    if @grouping.present?
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
-    end
 
+      if @assignment.enable_student_tests
+        if @assignment.non_regenerating_tokens
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:46]: It's great that you've simplified the logic; now, you can simplify it even further by combining this condition with the one above (i.e., `if @assignment.enable_student_tests && !@assignment.non_regenerating_tokens`). You don't need a separate branch to assign `@next_token_generation_time = nil` here because it'll be set in the "catch-all" statement on Line 57.

File: app/views/automated_tests/student_interface.html.erb [hunk 16]
```diff
@@ -31,6 +31,13 @@
     <% end %>
   </p>
 
+  <% unless @assignment.unlimited_tokens || @assignment.non_regenerating_tokens %>
```
Comment (david-yz-liu) [inline app/views/automated_tests/student_interface.html.erb:34]: Please move this to immediately below the "Tokens regenerate" line (i.e., above "Tokens available")

File: config/locales/en.yml [hunk 17]
```diff
@@ -1,5 +1,7 @@
 ---
 en:
+  automated_tests:
+    next_token_generation: Next Token Generation Time
```
Comment (david-yz-liu) [inline config/locales/en.yml:4]: 1. Only capitalize the first word (to make it consistent with the other labels on the autotesting page)
2. Move this string to a more specific locale file (see `automated_testing/views`). We're slowly moving strings out of this general file to better categorize them.

File: spec/controllers/automated_tests_controller_spec.rb [hunk 18]
```diff
@@ -2,6 +2,7 @@
   include AutomatedTestsHelper
 
   # TODO: add 'role is from a different course' shared tests to each route test below
+
```
Comment (david-yz-liu) [inline spec/controllers/automated_tests_controller_spec.rb:5]: Delete this line

File: spec/controllers/automated_tests_controller_spec.rb [hunk 19]
```diff
@@ -548,8 +549,83 @@
...
+      end
+
+      context 'when student tests are enabled and the student has a grouping' do
+        before do
+          get_as student, :student_interface, params: params
+        end
+
+        it 'should calculate the next token generation time' do
+          assignment.reload
+          current_time = Time.zone.now.change(hour: 14)
```
Comment (david-yz-liu) [inline spec/controllers/automated_tests_controller_spec.rb:578]: Good idea to change the time, but use the `travel` test helper: https://api.rubyonrails.org/v5.1.7/classes/ActiveSupport/Testing/TimeHelpers.html#method-i-travel.

File: spec/controllers/automated_tests_controller_spec.rb [hunk 20]
```diff
@@ -548,8 +549,83 @@
...
+          token_start_time = assignment.token_start_date
+          hours_since_start = [(current_time - token_start_time) / 1.hour, 0].max
```
Comment (david-yz-liu) [inline spec/controllers/automated_tests_controller_spec.rb:580]: It's good that this test is more specific now. However, because you've chosen specific numbers, you shouldn't just copy the code from the implementation, as that won't reveal bugs. Instead, you should separately figure out what time should be displayed, and use that in this test case.

In this case, based on the test configuration you've set up, if you travel to (e.g.) 14 hours in the future, the next token generation time should be 2 days from when the tokens started being available.

**Review by david-yz-liu — COMMENTED** [review]
@umututku03 looking good: the logic has been simplified and my previous comments addressed. I left a more detailed review, including UI-related items. After addressing these, please update the screenshot in the PR description as well. 👍

File: app/controllers/automated_tests_controller.rb [hunk 21]
```diff
@@ -35,14 +35,23 @@ def student_interface
...
 
-    unless @grouping.nil?
+    if @grouping.present?
       @grouping.refresh_test_tokens
       @authorized = flash_allowance(:notice,
                                     allowance_to(:run_tests?,
                                                  current_role,
                                                  context: { assignment: @assignment, grouping: @grouping })).value
+
+      if @assignment.enable_student_tests && !@assignment.non_regenerating_tokens
```
Comment (david-yz-liu) [inline app/controllers/automated_tests_controller.rb:45]: oh one more thing, should add a condition for `!@assignment.unlimited_tokens` here as well (matching what's in the view file). Otherwise I suppose it'd be possible for `@assignment.token_period` to be `nil` (or `0`), which wouldn't be good for the logic inside this block

**Review by david-yz-liu — APPROVED** [review]
@umututku03 great work on this! 🎉
